src/databricks/labs/ucx/workspace_access/groups.py

```diff
@@ -48,24 +47,67 @@ def get_by_workspace_group_name(self, workspace_group_name: str) -> MigrationGro
         else:
             return found[0]
 
+    def __len__(self):
+        return len(self.groups)
+
 
 class GroupManager:
-    SYSTEM_GROUPS: typing.ClassVar[list[str]] = ["users", "admins", "account users"]
-    SCIM_ATTRIBUTES = "id,displayName,meta,members"
+    _SYSTEM_GROUPS: typing.ClassVar[list[str]] = ["users", "admins", "account users"]
+    _SCIM_ATTRIBUTES = "id,displayName,meta,members"
 
     def __init__(self, ws: WorkspaceClient, groups: GroupsConfig):
         self._ws = ws
-        self.config = groups
+        # TODO: remove groups param in favor of _include_group_names and _backup_group_prefix
+        self._include_group_names = groups.selected
+        self._backup_group_prefix = groups.backup_group_prefix
         self._migration_state: GroupMigrationState = GroupMigrationState()
         self._account_groups = self._list_account_groups()
         self._workspace_groups = self._list_workspace_groups()
 
+    def prepare_groups_in_environment(self):
+        logger.info(
+            "Preparing groups in the current environment. At this step we'll verify that all groups "
+            "exist and are of the correct type. If some temporary groups are missing, they'll be created"
+        )
+        valid_group_names = self._get_valid_group_names_to_migrate()
+        self._set_migration_groups_with_backup(valid_group_names)
+        logger.info("Environment prepared successfully")
+
+    def has_groups(self) -> bool:
+        return len(self._migration_state.groups) > 0
+
+    @property
+    def migration_state(self) -> GroupMigrationState:
+        if len(self._migration_state) == 0:
```

Comment:
Why should a property care about logging the state length? we keep repeating the same message all over the places

Reply:
It's a question for @renardeinside  😜

Reply:
I would remove it from here.